File: frontend/utils/exerciseAnalyzer.ts
/**
 * Client-side exercise form analyzer.
 * Ported from backend/ml_models/pose_detection/exercise_analyzer.py.
 * Operates on MediaPipe pose landmarks (33 keypoints).
 */

export interface Point {
  x: number
  y: number
  z?: number
  visibility?: number
}

export type ExerciseKey =
  | 'squat'
  | 'pushup'
  | 'plank'
  | 'lunge'
  | 'bicep_curl'
  | 'shoulder_press'

export interface AnalysisResult {
  feedback: string[]
  score: number
  reps: number
  stage: 'up' | 'down' | null
}

// MediaPipe Pose landmark indices.
export const LM = {
  LEFT_SHOULDER: 11,
  RIGHT_SHOULDER: 12,
  LEFT_ELBOW: 13,
  RIGHT_ELBOW: 14,
  LEFT_WRIST: 15,
  RIGHT_WRIST: 16,
  LEFT_HIP: 23,
  RIGHT_HIP: 24,
  LEFT_KNEE: 25,
  RIGHT_KNEE: 26,
  LEFT_ANKLE: 27,
  RIGHT_ANKLE: 28,
} as const

const THRESHOLDS = {
  squat: { kneeDown: 90, kneeUp: 160, backMin: 160, backMax: 200 },
  pushup: { elbowDown: 90, elbowUp: 160, bodyMin: 160, bodyMax: 200 },
  plank: { bodyMin: 165, bodyMax: 195 },
  lunge: { kneeDown: 95, kneeUp: 160, backMin: 160, backMax: 200 },
  bicep_curl: { elbowFlexed: 50, elbowExtended: 160, elbowDriftMax: 0.08 },
  shoulder_press: { elbowDown: 95, elbowUp: 160, wristStackMax: 0.1 },
}

/**
 * Returns the angle (in degrees) at vertex b formed by points a-b-c.
 * 0 = collinear in same direction, 180 = straight line, 360 = full turn.
 */
export function angle(a: Point, b: Point, c: Point): number {
  const ab = { x: a.x - b.x, y: a.y - b.y }
  const cb = { x: c.x - b.x, y: c.y - b.y }
  const dot = ab.x * cb.x + ab.y * cb.y
  const magA = Math.hypot(ab.x, ab.y)
  const magC = Math.hypot(cb.x, cb.y)
  if (magA === 0 || magC === 0) return 0
  const cos = Math.max(-1, Math.min(1, dot / (magA * magC)))
  return (Math.acos(cos) * 180) / Math.PI
}

function mean(arr: number[]): number {
  if (arr.length === 0) return 0
  return arr.reduce((s, n) => s + n, 0) / arr.length
}

function getLm(landmarks: Point[] | null | undefined, idx: number): Point | null {
  if (!landmarks || !landmarks[idx]) return null
  return landmarks[idx]
}

export class ExerciseAnalyzer {
  private repCount = 0
  private stage: 'up' | 'down' | null = null
  private exercise: ExerciseKey = 'squat'

  setExercise(ex: ExerciseKey) {
    if (ex !== this.exercise) {
      this.exercise = ex
      this.reset()
    }
  }

  reset() {
    this.repCount = 0
    this.stage = null
  }

  getRepCount() {
    return this.repCount
  }

  analyze(landmarks: Point[] | null): AnalysisResult {
    if (this.exercise === 'squat') return this.analyzeSquat(landmarks)
    if (this.exercise === 'pushup') return this.analyzePushup(landmarks)
    if (this.exercise === 'plank') return this.analyzePlank(landmarks)
    if (this.exercise === 'lunge') return this.analyzeLunge(landmarks)
    if (this.exercise === 'bicep_curl') return this.analyzeBicepCurl(landmarks)
    return this.analyzeShoulderPress(landmarks)
  }

  private empty(): AnalysisResult {
    return { feedback: ['NO POSE DETECTED'], score: 0, reps: this.repCount, stage: this.stage }
  }

  private analyzeSquat(landmarks: Point[] | null): AnalysisResult {
    const hip = getLm(landmarks, LM.LEFT_HIP)
    const knee = getLm(landmarks, LM.LEFT_KNEE)
    const ankle = getLm(landmarks, LM.LEFT_ANKLE)
    const shoulder = getLm(landmarks, LM.LEFT_SHOULDER)
    if (!hip || !knee || !ankle || !shoulder) return this.empty()

    const kneeAngle = angle(hip, knee, ankle)
    const backAngle = angle(shoulder, hip, knee)

    const feedback: string[] = []
    const components: number[] = []

    if (kneeAngle < THRESHOLDS.squat.kneeDown) {
      if (this.stage !== 'down') this.stage = 'down'
      components.push(100)
    } else if (kneeAngle < 120) {
      components.push(70)
      feedback.push('GO DEEPER')
    } else {
      if (this.stage === 'down') {
        this.stage = 'up'
        this.repCount += 1
      }
      components.push(50)
    }

    if (backAngle > THRESHOLDS.squat.backMin && backAngle < THRESHOLDS.squat.backMax) {
      components.push(100)
    } else {
      components.push(50)
      feedback.push('KEEP BACK STRAIGHT')
    }

    if (Math.abs(knee.x - ankle.x) < 0.1) {
      components.push(100)
    } else {
      components.push(60)
      if (knee.x > ankle.x) feedback.push('KNEES TOO FORWARD')
    }

    return {
      feedback,
      score: Math.round(mean(components)),
      reps: this.repCount,
      stage: this.stage,
    }
  }

  private analyzePushup(landmarks: Point[] | null): AnalysisResult {
    const shoulder = getLm(landmarks, LM.LEFT_SHOULDER)
    const elbow = getLm(landmarks, LM.LEFT_ELBOW)
    const wrist = getLm(landmarks, LM.LEFT_WRIST)
    const hip = getLm(landmarks, LM.LEFT_HIP)
    const knee = getLm(landmarks, LM.LEFT_KNEE)
    if (!shoulder || !elbow || !wrist || !hip || !knee) return this.empty()

    const elbowAngle = angle(shoulder, elbow, wrist)
    const bodyAngle = angle(shoulder, hip, knee)

    const feedback: string[] = []
    const components: number[] = []

    if (elbowAngle < THRESHOLDS.pushup.elbowDown) {
      if (this.stage !== 'down') this.stage = 'down'
      components.push(100)
    } else if (elbowAngle < 120) {
      components.push(70)
      feedback.push('GO LOWER')
    } else {
      if (this.stage === 'down') {
        this.stage = 'up'
        this.repCount += 1
      }
      components.push(50)
    }

    if (bodyAngle > THRESHOLDS.pushup.bodyMin && bodyAngle < THRESHOLDS.pushup.bodyMax) {
      components.push(100)
    } else {
      components.push(50)
      if (bodyAngle < THRESHOLDS.pushup.bodyMin) feedback.push('HIPS TOO LOW')
      else feedback.push('HIPS TOO HIGH')
    }

    return {
      feedback,
      score: Math.round(mean(components)),
      reps: this.repCount,
      stage: this.stage,
    }
  }

  private analyzePlank(landmarks: Point[] | null): AnalysisResult {
    const shoulder = getLm(landmarks, LM.LEFT_SHOULDER)
    const hip = getLm(landmarks, LM.LEFT_HIP)
    const ankle = getLm(landmarks, LM.LEFT_ANKLE)
    if (!shoulder || !hip || !ankle) return this.empty()

    const bodyAngle = angle(shoulder, hip, ankle)
    const feedback: string[] = []
    const components: number[] = []

    if (bodyAngle > THRESHOLDS.plank.bodyMin && bodyAngle < THRESHOLDS.plank.bodyMax) {
      components.push(100)
    } else {
      components.push(50)
      if (bodyAngle < THRESHOLDS.plank.bodyMin) feedback.push('HIPS TOO LOW')
      else feedback.push('HIPS TOO HIGH')
    }

    return {
      feedback,
      score: Math.round(mean(components)),
      reps: this.repCount,
      stage: null,
    }
  }

  private analyzeLunge(landmarks: Point[] | null): AnalysisResult {
    const hip = getLm(landmarks, LM.LEFT_HIP)
    const knee = getLm(landmarks, LM.LEFT_KNEE)
    const ankle = getLm(landmarks, LM.LEFT_ANKLE)
    const shoulder = getLm(landmarks, LM.LEFT_SHOULDER)
    if (!hip || !knee || !ankle || !shoulder) return this.empty()

    const kneeAngle = angle(hip, knee, ankle)
    const backAngle = angle(shoulder, hip, knee)

    const feedback: string[] = []
    const components: number[] = []

    if (kneeAngle < THRESHOLDS.lunge.kneeDown) {
      if (this.stage !== 'down') this.stage = 'down'
      components.push(100)
    } else if (kneeAngle < 120) {
      components.push(70)
      feedback.push('DROP DEEPER')
    } else {
      if (this.stage === 'down') {
        this.stage = 'up'
        this.repCount += 1
      }
      components.push(50)
    }

    if (backAngle > THRESHOLDS.lunge.backMin && backAngle < THRESHOLDS.lunge.backMax) {
      components.push(100)
    } else {
      components.push(50)
      feedback.push('KEEP TORSO UPRIGHT')
    }

    if (knee.x - ankle.x < 0.1) {
      components.push(100)
    } else {
      components.push(60)
      feedback.push('FRONT KNEE PAST TOE')
    }

    return {
      feedback,
      score: Math.round(mean(components)),
      reps: this.repCount,
      stage: this.stage,
    }
  }

  private analyzeBicepCurl(landmarks: Point[] | null): AnalysisResult {
    const shoulder = getLm(landmarks, LM.LEFT_SHOULDER)
    const elbow = getLm(landmarks, LM.LEFT_ELBOW)
    const wrist = getLm(landmarks, LM.LEFT_WRIST)
    if (!shoulder || !elbow || !wrist) return this.empty()

    const elbowAngle = angle(shoulder, elbow, wrist)

    const feedback: string[] = []
    const components: number[] = []

    if (elbowAngle < THRESHOLDS.bicep_curl.elbowFlexed) {
      if (this.stage !== 'down') this.stage = 'down'
      components.push(100)
    } else if (elbowAngle < 110) {
      components.push(70)
      feedback.push('CURL HIGHER')
    } else if (elbowAngle > THRESHOLDS.bicep_curl.elbowExtended) {
      if (this.stage === 'down') {
        this.stage = 'up'
        this.repCount += 1
      }
      components.push(100)
    } else {
      components.push(70)
    }

    if (Math.abs(elbow.x - shoulder.x) < THRESHOLDS.bicep_curl.elbowDriftMax) {
      components.push(100)
    } else {
      components.push(60)
      feedback.push('KEEP ELBOW STILL')
    }

    return {
      feedback,
      score: Math.round(mean(components)),
      reps: this.repCount,
      stage: this.stage,
    }
  }

  private analyzeShoulderPress(landmarks: Point[] | null): AnalysisResult {
    const shoulder = getLm(landmarks, LM.LEFT_SHOULDER)
    const elbow = getLm(landmarks, LM.LEFT_ELBOW)
    const wrist = getLm(landmarks, LM.LEFT_WRIST)
    if (!shoulder || !elbow || !wrist) return this.empty()

    const elbowAngle = angle(shoulder, elbow, wrist)
    // MediaPipe y is top-down: smaller y = higher on screen.
    const wristAboveShoulder = wrist.y < shoulder.y

    const feedback: string[] = []
    const components: number[] = []

    if (elbowAngle < THRESHOLDS.shoulder_press.elbowDown && !wristAboveShoulder) {
      if (this.stage !== 'down') this.stage = 'down'
      components.push(100)
    } else if (elbowAngle > THRESHOLDS.shoulder_press.elbowUp && wristAboveShoulder) {
      if (this.stage === 'down') {
        this.stage = 'up'
        this.repCount += 1
      }
      components.push(100)
    } else {
      components.push(70)
      if (elbowAngle > THRESHOLDS.shoulder_press.elbowUp && !wristAboveShoulder) {
        feedback.push('PRESS HIGHER')
      }
    }

    if (Math.abs(wrist.x - elbow.x) < THRESHOLDS.shoulder_press.wristStackMax) {
      components.push(100)
    } else {
      components.push(60)
      feedback.push('STACK WRIST OVER ELBOW')
    }

    return {
      feedback,
      score: Math.round(mean(components)),
      reps: this.repCount,
      stage: this.stage,
    }
  }
}

/**
 * MediaPipe pose connections — edges between landmark indices used to draw the skeleton.
 */
export const POSE_CONNECTIONS: [number, number][] = [
  [11, 12], // shoulders
  [11, 13],
  [13, 15], // left arm
  [12, 14],
  [14, 16], // right arm
  [11, 23],
  [12, 24], // shoulder to hip
  [23, 24], // hips
  [23, 25],
  [25, 27], // left leg
  [24, 26],
  [26, 28], // right leg
]

File: ml_models/pose_detection/exercise_analyzer.py
"""
Exercise Form Analyzer
Analyzes specific exercises and provides real-time feedback
"""

import numpy as np
import mediapipe as mp
try:
    from pose_detector import PoseDetector
except ImportError:
    from .pose_detector import PoseDetector


class ExerciseAnalyzer:
    """
    Analyzes different exercise types and provides form correction feedback
    """
    
    def __init__(self):
        """
        Initialize exercise templates and thresholds
        """
        self.pose_detector = PoseDetector()
        self.mp_pose = mp.solutions.pose
        
        # Rep counting state
        self.rep_count = 0
        self.stage = None  # "up" or "down"
        
        # Exercise thresholds
        self.thresholds = {
            'squat': {
                'knee_angle_down': 90,
                'knee_angle_up': 160,
                'back_angle_min': 160,
                'back_angle_max': 200
            },
            'pushup': {
                'elbow_angle_down': 90,
                'elbow_angle_up': 160,
                'body_angle_min': 160,
                'body_angle_max': 200
            },
            'plank': {
                'body_angle_min': 165,
                'body_angle_max': 195,
                'min_hold_time': 5  # seconds
            },
            'lunge': {
                'knee_angle_down': 95,
                'knee_angle_up': 160,
                'back_angle_min': 160,
                'back_angle_max': 200
            },
            'bicep_curl': {
                'elbow_flexed': 50,
                'elbow_extended': 160,
                'elbow_drift_max': 0.08
            },
            'shoulder_press': {
                'elbow_down': 95,
                'elbow_up': 160,
                'wrist_stack_max': 0.1
            }
        }
    
    def analyze_squat(self, landmarks):
        """
        Analyze squat form
        
        Args:
            landmarks: MediaPipe pose landmarks
            
        Returns:
            dict: {
                'feedback': List of correction messages,
                'score': Form accuracy score (0-100),
                'angles': Dict of joint angles,
                'stage': Current movement stage
            }
        """
        if not landmarks:
            return {'feedback': ['No pose detected'], 'score': 0, 'angles': {}, 'stage': None}
        
        # Get joint coordinates
        hip = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_HIP.value)
        knee = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_KNEE.value)
        ankle = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_ANKLE.value)
        shoulder = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_SHOULDER.value)
        
        # Calculate angles
        knee_angle = self.pose_detector.calculate_angle(hip, knee, ankle)
        back_angle = self.pose_detector.calculate_angle(shoulder, hip, knee)
        
        feedback = []
        score_components = []
        
        # Depth check
        if knee_angle < self.thresholds['squat']['knee_angle_down']:
            if self.stage != "down":
                self.stage = "down"
            score_components.append(100)
        elif knee_angle < 120:
            score_components.append(70)
            feedback.append("Go deeper")
        else:
            if self.stage == "down":
                self.stage = "up"
                self.rep_count += 1
            score_components.append(50)
        
        # Back straightness
        if self.thresholds['squat']['back_angle_min'] < back_angle < self.thresholds['squat']['back_angle_max']:
            score_components.append(100)
        else:
            score_components.append(50)
            feedback.append("Keep back straight")
        
        # Knee alignment
        knee_x = knee[0]
        ankle_x = ankle[0]
        if abs(knee_x - ankle_x) < 0.1:
            score_components.append(100)
        else:
            score_components.append(60)
            if knee_x > ankle_x:
                feedback.append("Knees too forward")
        
        final_score = int(np.mean(score_components)) if score_components else 0
        
        return {
            'feedback': feedback,
            'score': final_score,
            'angles': {'knee': knee_angle, 'back': back_angle},
            'stage': self.stage,
            'reps': self.rep_count
        }
    
    def analyze_pushup(self, landmarks):
        """
        Analyze push-up form
        
        Args:
            landmarks: MediaPipe pose landmarks
            
        Returns:
            dict: {
                'feedback': List of correction messages,
                'score': Form accuracy score (0-100),
                'angles': Dict of joint angles,
                'stage': Current movement stage
            }
        """
        if not landmarks:
            return {'feedback': ['No pose detected'], 'score': 0, 'angles': {}, 'stage': None}
        
        # Get joint coordinates
        shoulder = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_SHOULDER.value)
        elbow = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_ELBOW.value)
        wrist = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_WRIST.value)
        hip = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_HIP.value)
        knee = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_KNEE.value)
        
        # Calculate angles
        elbow_angle = self.pose_detector.calculate_angle(shoulder, elbow, wrist)
        body_angle = self.pose_detector.calculate_angle(shoulder, hip, knee)
        
        feedback = []
        score_components = []
        
        # Depth check
        if elbow_angle < self.thresholds['pushup']['elbow_angle_down']:
            if self.stage != "down":
                self.stage = "down"
            score_components.append(100)
        elif elbow_angle < 120:
            score_components.append(70)
            feedback.append("Go lower")
        else:
            if self.stage == "down":
                self.stage = "up"
                self.rep_count += 1
            score_components.append(50)
        
        # Body alignment
        if self.thresholds['pushup']['body_angle_min'] < body_angle < self.thresholds['pushup']['body_angle_max']:
            score_components.append(100)
        else:
            score_components.append(50)
            if body_angle < self.thresholds['pushup']['body_angle_min']:
                feedback.append("Hips too low")
            else:
                feedback.append("Hips too high")
        
        final_score = int(np.mean(score_components)) if score_components else 0
        
        return {
            'feedback': feedback,
            'score': final_score,
            'angles': {'elbow': elbow_angle, 'body': body_angle},
            'stage': self.stage,
            'reps': self.rep_count
        }
    
    def analyze_plank(self, landmarks):
        """
        Analyze plank form
        
        Args:
            landmarks: MediaPipe pose landmarks
            
        Returns:
            dict: {
                'feedback': List of correction messages,
                'score': Form accuracy score (0-100),
                'angles': Dict of joint angles
            }
        """
        if not landmarks:
            return {'feedback': ['No pose detected'], 'score': 0, 'angles': {}}
        
        # Get joint coordinates
        shoulder = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_SHOULDER.value)
        hip = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_HIP.value)
        ankle = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_ANKLE.value)
        
        # Calculate body alignment angle
        body_angle = self.pose_detector.calculate_angle(shoulder, hip, ankle)
        
        feedback = []
        score_components = []
        
        # Body straightness
        if self.thresholds['plank']['body_angle_min'] < body_angle < self.thresholds['plank']['body_angle_max']:
            score_components.append(100)
        else:
            score_components.append(50)
            if body_angle < self.thresholds['plank']['body_angle_min']:
                feedback.append("Hips too low")
            else:
                feedback.append("Hips too high")
        
        final_score = int(np.mean(score_components)) if score_components else 0
        
        return {
            'feedback': feedback,
            'score': final_score,
            'angles': {'body': body_angle}
        }
    
    def analyze_lunge(self, landmarks):
        """
        Analyze lunge form (front-leg view, mirrors squat state machine)

        Args:
            landmarks: MediaPipe pose landmarks

        Returns:
            dict: {
                'feedback': List of correction messages,
                'score': Form accuracy score (0-100),
                'angles': Dict of joint angles,
                'stage': Current movement stage,
                'reps': Current rep count
            }
        """
        if not landmarks:
            return {'feedback': ['No pose detected'], 'score': 0, 'angles': {}, 'stage': None}

        hip = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_HIP.value)
        knee = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_KNEE.value)
        ankle = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_ANKLE.value)
        shoulder = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_SHOULDER.value)

        knee_angle = self.pose_detector.calculate_angle(hip, knee, ankle)
        back_angle = self.pose_detector.calculate_angle(shoulder, hip, knee)

        feedback = []
        score_components = []

        # Depth check (front leg)
        if knee_angle < self.thresholds['lunge']['knee_angle_down']:
            if self.stage != "down":
                self.stage = "down"
            score_components.append(100)
        elif knee_angle < 120:
            score_components.append(70)
            feedback.append("Drop deeper")
        else:
            if self.stage == "down":
                self.stage = "up"
                self.rep_count += 1
            score_components.append(50)

        # Torso uprightness
        if self.thresholds['lunge']['back_angle_min'] < back_angle < self.thresholds['lunge']['back_angle_max']:
            score_components.append(100)
        else:
            score_components.append(50)
            feedback.append("Keep torso upright")

        # Front knee tracking (knee should not drift far past ankle)
        if (knee[0] - ankle[0]) < 0.1:
            score_components.append(100)
        else:
            score_components.append(60)
            feedback.append("Front knee past toe")

        final_score = int(np.mean(score_components)) if score_components else 0

        return {
            'feedback': feedback,
            'score': final_score,
            'angles': {'knee': knee_angle, 'back': back_angle},
            'stage': self.stage,
            'reps': self.rep_count
        }

    def analyze_bicep_curl(self, landmarks):
        """
        Analyze bicep curl form using a single elbow angle.

        Args:
            landmarks: MediaPipe pose landmarks

        Returns:
            dict: {
                'feedback': List of correction messages,
                'score': Form accuracy score (0-100),
                'angles': Dict of joint angles,
                'stage': Current movement stage,
                'reps': Current rep count
            }
        """
        if not landmarks:
            return {'feedback': ['No pose detected'], 'score': 0, 'angles': {}, 'stage': None}

        shoulder = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_SHOULDER.value)
        elbow = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_ELBOW.value)
        wrist = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_WRIST.value)

        elbow_angle = self.pose_detector.calculate_angle(shoulder, elbow, wrist)

        feedback = []
        score_components = []

        # Stage 'down' = curled at top (small elbow angle); 'up' = arm extended.
        # Counts on the down -> up (curl-and-lower) transition, like squat.
        if elbow_angle < self.thresholds['bicep_curl']['elbow_flexed']:
            if self.stage != "down":
                self.stage = "down"
            score_components.append(100)
        elif elbow_angle < 110:
            score_components.append(70)
            feedback.append("Curl higher")
        elif elbow_angle > self.thresholds['bicep_curl']['elbow_extended']:
            if self.stage == "down":
                self.stage = "up"
                self.rep_count += 1
            score_components.append(100)
        else:
            score_components.append(70)

        # Elbow stability — shouldn't drift forward/back from shoulder line.
        if abs(elbow[0] - shoulder[0]) < self.thresholds['bicep_curl']['elbow_drift_max']:
            score_components.append(100)
        else:
            score_components.append(60)
            feedback.append("Keep elbow still")

        final_score = int(np.mean(score_components)) if score_components else 0

        return {
            'feedback': feedback,
            'score': final_score,
            'angles': {'elbow': elbow_angle},
            'stage': self.stage,
            'reps': self.rep_count
        }

    def analyze_shoulder_press(self, landmarks):
        """
        Analyze overhead shoulder press form.

        Args:
            landmarks: MediaPipe pose landmarks

        Returns:
            dict: {
                'feedback': List of correction messages,
                'score': Form accuracy score (0-100),
                'angles': Dict of joint angles,
                'stage': Current movement stage,
                'reps': Current rep count
            }
        """
        if not landmarks:
            return {'feedback': ['No pose detected'], 'score': 0, 'angles': {}, 'stage': None}

        shoulder = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_SHOULDER.value)
        elbow = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_ELBOW.value)
        wrist = self.pose_detector.get_landmark_coords(landmarks, self.mp_pose.PoseLandmark.LEFT_WRIST.value)

        elbow_angle = self.pose_detector.calculate_angle(shoulder, elbow, wrist)

        feedback = []
        score_components = []

        # MediaPipe y is top-down: smaller y = higher on screen.
        wrist_above_shoulder = wrist[1] < shoulder[1]

        # Stage 'down' = rack position (bent elbows, wrists at/below shoulder line).
        # Stage 'up'   = locked overhead (extended elbows, wrists above shoulders).
        if elbow_angle < self.thresholds['shoulder_press']['elbow_down'] and not wrist_above_shoulder:
            if self.stage != "down":
                self.stage = "down"
            score_components.append(100)
        elif elbow_angle > self.thresholds['shoulder_press']['elbow_up'] and wrist_above_shoulder:
            if self.stage == "down":
                self.stage = "up"
                self.rep_count += 1
            score_components.append(100)
        else:
            score_components.append(70)
            if elbow_angle > self.thresholds['shoulder_press']['elbow_up'] and not wrist_above_shoulder:
                feedback.append("Press higher")

        # Wrist should stack roughly over the elbow at lockout.
        if abs(wrist[0] - elbow[0]) < self.thresholds['shoulder_press']['wrist_stack_max']:
            score_components.append(100)
        else:
            score_components.append(60)
            feedback.append("Stack wrist over elbow")

        final_score = int(np.mean(score_components)) if score_components else 0

        return {
            'feedback': feedback,
            'score': final_score,
            'angles': {'elbow': elbow_angle},
            'stage': self.stage,
            'reps': self.rep_count
        }

    def reset_reps(self):
        """Reset rep counter and stage"""
        self.rep_count = 0
        self.stage = None
    
    def get_rep_count(self):
        """Get current rep count"""
        return self.rep_count
